### ISSUE TITLE
Using Kontena volumes fails with Docker v2 volume plugins

### ISSUE THREAD

**Issue by nevalla** [kontena/kontena#2261]
When using Docker volume plugins, Docker uses version/tag in volume driver information
```
$ docker volume inspect s3fs-test
[
    {
        "Driver": "rexray/s3fs:latest",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/plugins/165a4e198e4ff02d1a3f1167cdfe132e046efa45db6a2384e03fdfd4f6e9a381/rootfs",
        "Name": "s3fs-test",
        "Options": {},
        "Scope": "global",
        "Status": {
            "availabilityZone": "",
            "fields": null,
            "iops": 0,
            "name": "s3fs-test",
            "server": "s3fs",
            "service": "s3fs",
            "size": 0,
            "type": ""
        }
    }
]
```

When deploying a stack, `Kontena::Workers::Volumes::VolumeManager` checks exisiting volumes without the version information and the comparison will fail causing:

```
GridServiceInstanceDeployer::ServiceError: Kontena::Workers::Volumes::VolumeManager::DriverMismatchError: Volume driver not as expected. Expected rexray/s3fs, existing volume has rexray/s3fs:latest
```

If you try to create volume with version information `kontena volume create --driver rexray/s3fs:latest ...` Kontena scheduler cannot find a proper node because version/tag information is not used in the comparison.

**Comment by SpComb** [comment]
This is a bad interaction between #2061 and #2135. In #2061 we decided to explicitly split out the version and only use the driver name, and then in #2135 we only tested it with the un-versioned `local` driver.

**Comment by jakolehm** [comment]
@nevalla did you install driver as a docker plugin? If yes, workaround might be to set `--alias` at install.

**Comment by SpComb** [comment]
No fix for 1.3 yet, bumping to 1.4. Might also be to a 1.3.2 if we want to.

**Comment by jakolehm** [comment]
Impossible to get v2 volume plugins to work because agent throws `DriverMismatchError` (even with `--alias`).

Tested with CoreOS Alpha & rexray/s3fs.

```
$ docker version
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.6
 Git commit:   89658be
 Built:        Wed Jun  7 23:07:34 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.6
 Git commit:   89658be
 Built:        Wed Jun  7 23:07:34 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by Edke** [comment]
Also spent few hours to get this to work, my notes:

- target nodes pre-installed with driver as kontena still does not distribute plugins across grid (#2320)
- installed using --alias, version of rexray/s3fs 0.9.1 has permissions issue as also noted on https://hub.docker.com/r/rexray/s3fs/, therefore 0.9.0 version is used  
```
$ docker plugin ls 
ID                  NAME                 DESCRIPTION               ENABLED
06dd4c96420f        rexray/s3fs:latest   REX-Ray for Amazon S3FS   true
```

### a) creating kontena volume with `:latest` tag

```
$ kontena volume create  --scope grid --driver rexray/s3fs:latest adptw-test2-vol
$ kontena volume ls 
NAME              SCOPE   DRIVER               CREATED_AT
adptw-test2-vol   grid    rexray/s3fs:latest   12 minutes ago
```

Deploy fails as plugin does not match node's plugins:

```
$ kontena stack ls 
NAME  STACK  SERVICES_COUNT  STATE  PORTS
kraken@thingol:~/sandbox/stack-multiple-nodes                                                                                                                                       [09:24:21]$ kontena stack install
 [done] Creating stack test1      
 [done] Triggering deployment of stack test1     
 [fail] Waiting for deployment to start     
        Cannot find applicable node for service instance production/test1/web1-1: Did not find any nodes with required volume drivers: ["rexray/s3fs:latest"]
 [error] RuntimeError : deploy failed
         See /home/kraken/.kontena/kontena.log or run the command again with environment DEBUG=true set to see the full exception
```

Target node info:
```
$ kontena node show smaug
smaug:
  id: VJDW:5QRW:BU45:ZMMB:4AZI:4RIT:LC2T:6DY6:QCE4:2EA7:6CQU:4GNE
  agent version: 1.3.0
  docker version: 17.05.0-ce
  connected: yes
  last connect: 2017-06-15T21:16:10.465+00:00
  last seen: 2017-06-23T07:26:40.177Z
  public ip: ***
  private ip: ***
  overlay ip: 10.81.0.1
  os: Arch Linux
  kernel: 4.11.5-1-ARCH
  drivers:
    storage: overlay2
    volume: rexray/s3fs,local
  initial node: yes
  labels:
  stats:
    cpus: 2
    load: 1.32 0.51 0.26
    memory: 0.52 of 1.96 GB
    filesystem:
      - /var/lib/docker: 26.91 of 39.34 GB
```

### b) creating kontena volume without `:latest` tag

```
$ kontena volume create  --scope grid --driver rexray/s3fs adptw-test2-vol
$ kontena volume ls 
NAME              SCOPE   DRIVER        CREATED_AT
adptw-test2-vol   grid    rexray/s3fs   3 seconds ago
```

DriverMismatchError:

```
$ kontena stack install
 [done] Creating stack test1      
 [done] Triggering deployment of stack test1     
 [done] Waiting for deployment to start     
 [fail] Deploying service web1     
Deployment of service production/test1/web1 failed:
  - halting deploy of production/test1/web1, one or more instances failed
  - GridServiceInstanceDeployer::ServiceError: Kontena::Workers::Volumes::VolumeManager::DriverMismatchError: Volume driver not as expected. Expected rexray/s3fs, existing volume has rexray/s3fs:latest (on node smaug)
```

**Comment by SpComb** [comment]
Alright, doing some local testing with CoreOS alpha, Docker 17.05 to try and understand how these plugin / volume driver names and versions work re  #2518 and #2526

> The "unqualified" driver only works for either alias or latest version.

Right, it seems like the volume driver versions behave as the Docker image versions: ` vieux/sshfs` without a suffix just means `vieux/sshfs:latest`, both for `docker volume create --driver` and `docker plugin *`.

However, the Docker plugins have the added confusion of having a local suffix for the plugin name which is independent of the actual plugin version suffix, as per `docker plugin install --alias` (custom local name suffix that is different from the plugin version suffix) and `docker plugin upgrade` (keep the local name suffix, but upgrade the plugin version). This can be seen in the `docker plugin inspect` `Name` (taken from `install --alias`, stays the same on `upgrade`) vs `PluginReference` (taken from the actual `install` remote, changes on `upgrade`).

Additional it seems like Docker Hub plugins have similar updatable tags as images... the remote `vieux:lastest` plugin reference can be updated with a new plugin version? I assume this based on the default `docker plugin upgrade vieux/sshfs` behavior: `Upgrading plugin vieux/sshfs:latest from vieux/sshfs:latest to vieux/sshfs:latest`

**Comment by SpComb** [comment]
OFC I also saw the Docker API doing some crazy stuff when testing...

Some commands run earlier to install plugins / create volumes:
```
$ docker volume create --driver vieux/sshfs --name sshfs-test -o sshcmd=core@localhost:/tmp^C
$ docker volume ls
DRIVER               VOLUME NAME
vieux/sshfs:latest   ssh-test1
vieux/sshfs:latest   ssh-test2
vieux/sshfs:latest   sshfs-test
$ docker plugin install vieux/sshfs:1.0
$ docker plugin install --alias vieux/sshfs:test vieux/sshfs:1.0
```

And then:
```
$ docker plugin ls
ID                  NAME                 DESCRIPTION               ENABLED
79b9052f8801        vieux/sshfs:latest   sshFS plugin for Docker   true
95ba6f52efb5        vieux/sshfs:1.0      sshFS plugin for Docker   true
a74005c94ffc        vieux/sshfs:test     sshFS plugin for Docker   true
$ docker plugin disable vieux/sshfs:latest
Error response from daemon: plugin vieux/sshfs:latest is in use
$ docker volume ls
vieux/sshfs:latest   ssh-test1
vieux/sshfs:latest   ssh-test2
vieux/sshfs:latest   sshfs-test
$ docker volume rm sshfs-test
sshfs-test
$ docker volume rm ssh-test1 
ssh-test1
$ docker volume rm ssh-test2
ssh-test2
$ docker plugin disable vieux/sshfs:latest
Error response from daemon: plugin vieux/sshfs:latest is in use
$ docker volume ls
DRIVER              VOLUME NAME
vieux/sshfs:1.0     ssh-test1
vieux/sshfs:test    ssh-test1
vieux/sshfs:1.0     ssh-test2
vieux/sshfs:test    ssh-test2
vieux/sshfs:1.0     sshfs-test
vieux/sshfs:test    sshfs-test
```

No idea how it picked up multiple identically named copies of the same volume, one for each of the other installed plugin versions... I imagine this would confuse the agent.

**Comment by jnummelin** [comment]
> I imagine this would confuse the agent.
Quite possibly. It also seems to confuse docker itself:
```
$ docker run --rm -ti -v wordpress.wp-s3fs:/data alpine sh
docker: Error response from daemon: error while mounting volume '/var/lib/docker/plugins/f0d5c0da35f2ce3beb2f692eb37c6ecef56fef9bbb040bbea645e795ff046275/rootfs': VolumeDriver.Mount: {"Error":"error executing xcli"}.

```
That's from a case where I had multiple versions of `rexray/s3fs` installed and had multiple identically named volumes.

**Comment by SpComb** [comment]
> `VolumeDriver.Mount: {"Error":"error executing xcli"}.`

That looks like the `rexray/s3fs:0.9.1` permissions problem that @Edke also mentioned: https://github.com/codedellemc/rexray/issues/891

**Comment by SpComb** [comment]
Documenting the design discussion on fixing this issue of dealing with the Docker plugin / volume driver versions for configuring the `--driver` to use for Kontena volumes.

> I think we need to remove volume driver version specify. It's too confusing to users and also freezes plugin version (hard to install newer versions to fresh nodes without removing old installation first).

My understanding of the issue with driver versions here is something like the following:

* node-1 `docker plugin install rexray/s3fs:0.8.2`
* node-1 `docker plugin upgrade rexray/s3fs:0.8.2 rexray/s3fs:0.9.1`
* node-2 `docker plugin install rexray/s3fs:0.9.1`

This results in the both nodes running the same 0.9.1 version, but the Docker volume driver on node-1 is still `rexray/s3fs:0.8.2`, whereas on node-2 it is `rexray/s3fs:0.9.1`. There would be no way to create/deploy the same Kontena volume to both nodes in this case.

The two approaches to avoiding this kind of foot-gun situation:

* #2526 assuming that nodes only have a single plugin version installed, and all versions of the same plugin are compatible with eachother across nodes, have Kontena automatically use the installed Docker plugin version for each node

    This will work in the above nightmare scenario of using `docker plugin install rexray/s3fs:x.y.z` and ending up with different volume driver versions on different nodes.

* #2518 allow specifying an explicit `kontena volume create --driver rexray/s3fs:stable` version, or falling back to the implicit Docker `rexray/s3fs:latest` default for `kontena volume create --driver rexray/s3fs`

    For the typical case of `docker plugin install vieux/sshfs`, which results in the same `vieux/sshfs:latest` plugin being installed on all nodes, this would work fine with `kontena volume create --driver vieux/sshfs`. You can use `docker plugin upgrade vieux/sshfs` on each node in turn to upgrade the plugin version without affecting the volume drivers used.

    If you need to install specific versions of plugins, then it is a requirement to use custom local plugin versions using something like `docker plugin install --alias rexray/s3fs:stable rexray/s3fs:0.8.2`, to ensure that the plugin name is the same across nodes and versioned installs/upgrades. The local alias can then be used per `kontena volume create --driver rexray/s3fs:stable`.

    You could also use upstream labels like `docker plugin install rexray/s3fs:edge`, or use multiple different local plugin aliases to install different (possibly incompatible?) versions of the same plugin across nodes.

The two approaches behave identically in the typical usecase of `docker plugin install vieux/sshfs`, and both approaches will have `kontena volume create --driver vieux/sshfs` use the installed `vieux/sshfs:latest` plugin.

**Comment by jnummelin** [comment]
fixed in #2526
